File: core/student-mistake-database/src/main/java/com/tingyun/smartmistakebook/core/student/mistake/database/StudentReviewResponseBindingIssuer.java
package com.tingyun.smartmistakebook.core.student.mistake.database;

import java.util.Objects;

/** Owner-only response binding; neither the signer nor its dynamic issue method is public JVM ABI. */
abstract class StudentReviewResponseBindingIssuer {
    abstract String algorithmVersion();

    abstract String issue(String scopeCanonicalFingerprint, String canonicalResponse);
}

/** Uses the generation- and relay-epoch-bound student outbox key under a separate MAC domain. */
final class StudentOutboxBoundReviewResponseBindingIssuer
        extends StudentReviewResponseBindingIssuer {
    static final String ALGORITHM_VERSION =
            "student-outbox-keyed-review-response-hmac-sha256-v2";

    private final StudentOutboxAuthenticator authenticator;

    StudentOutboxBoundReviewResponseBindingIssuer(StudentOutboxAuthenticator authenticator) {
        this.authenticator = Objects.requireNonNull(authenticator, "authenticator");
    }

    @Override
    String algorithmVersion() {
        return ALGORITHM_VERSION;
    }

    @Override
    String issue(String scopeCanonicalFingerprint, String canonicalResponse) {
        return authenticator.issueReviewResponseBinding(
                scopeCanonicalFingerprint, canonicalResponse);
    }
}

File: core/student-mistake-database/src/main/kotlin/com/tingyun/smartmistakebook/core/student/mistake/database/StudentReviewResponseBindingIssuer.kt
package com.tingyun.smartmistakebook.core.student.mistake.database

import com.tingyun.smartmistakebook.core.model.CanonicalSha256
import com.tingyun.smartmistakebook.core.model.storage.ReviewResponseForm
import com.tingyun.smartmistakebook.core.model.storage.StudentProblemRevisionRef

internal fun studentReviewResponseBindingScopeFingerprint(
    learnerId: String,
    problemRevision: StudentProblemRevisionRef,
    reviewSessionId: String,
    reviewQueueItemId: String,
    presentationId: String,
    attemptOrdinal: Int,
    submissionId: String,
    responseForm: ReviewResponseForm,
): String =
    CanonicalSha256("student-review-response-binding-scope-v1")
        .field("learnerId", learnerId)
        .field("problemRevision", problemRevision.canonicalFingerprint)
        .field("reviewSessionId", reviewSessionId)
        .field("reviewQueueItemId", reviewQueueItemId)
        .field("presentationId", presentationId)
        .field("attemptOrdinal", attemptOrdinal)
        .field("submissionId", submissionId)
        .field("responseForm", responseForm.name)
        .finish()
